### PR TITLE
Allow mobile node to be restarted

### PR DESCRIPTION
**Pull request by cpacia**
This PR adds a restart method to the mobile node.

### CODE REVIEW

File: api/gateway.go [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
 import (
-	"net"
-	"net/http"
-
 	"github.com/OpenBazaar/openbazaar-go/core"
 	"github.com/OpenBazaar/openbazaar-go/schema"
 	"github.com/ipfs/go-ipfs/core/corehttp"
 	"github.com/op/go-logging"
+	"net"
+	"net/http"
```
Comment (placer14) [inline api/gateway.go]: Can you remove this no-op diff?

File: mobile/cmd/main.go [hunk 2]
```diff
@@ -2,11 +2,11 @@ package main
...
 import (
 	"fmt"
-	"os"
-	"sync"
-
 	"github.com/OpenBazaar/openbazaar-go/mobile"
 	"github.com/jessevdk/go-flags"
+	"os"
+	"sync"
+	"time"
```
Comment (placer14) [inline mobile/cmd/main.go]: Let's keep the packages organized consistently with newlines between each group:
```
import (
  // stdlib_packages

  // external_packages
 
  // gx_packages
)
```

File: mobile/node.go [hunk 3]
```diff
@@ -295,7 +297,9 @@ func NewNodeWithConfig(config *NodeConfig, password string, mnemonic string) (*N
 	ncfg := ipfs.PrepareIPFSConfig(r, ignoredURI, config.Testnet, config.Testnet)
 	ncfg.Routing = constructMobileRouting
 
-	return &Node{OpenBazaarNode: core.Node, config: *config, ipfsConfig: ncfg, apiConfig: apiConfig}, nil
+	core.Node.PublishLock.Lock()
+
+	return &Node{OpenBazaarNode: core.Node, config: *config, ipfsConfig: ncfg, apiConfig: apiConfig, mtx: sync.Mutex{}}, nil
```
Comment (placer14) [inline mobile/node.go]: Can we avoid using the global node and only use this Node instance? I'm concerned we're using a global object state which does not match this instance state? (Why do we have a global `core.Node` anyway?)

File: mobile/node.go [hunk 4]
```diff
@@ -338,10 +342,18 @@ func (n *Node) startIPFSNode(repoPath string, config *ipfscore.BuildCfg) (*ipfsc
...
+
+	return n.start()
+}
+
+func (n *Node) start() error {
 	nd, ctx, err := n.startIPFSNode(n.config.RepoPath, n.ipfsConfig)
 	if err != nil {
 		return err
 	}
+	obNode := n.OpenBazaarNode
```
Comment (placer14) [inline mobile/node.go]: This change doesn't appear to do anything for us? Can we revert?

File: mobile/node.go [hunk 5]
```diff
@@ -64,6 +65,9 @@ type Node struct {
 	cancel         context.CancelFunc
 	ipfsConfig     *ipfscore.BuildCfg
 	apiConfig      *apiSchema.APIConfig
+	gateway        *api.Gateway
+	started        bool
+	mtx            sync.Mutex
```
Comment (placer14) [inline mobile/node.go]: Can you make the fieldname more descriptive? `startupMutex` or similar?

File: mobile/node.go [hunk 6]
```diff
@@ -404,78 +416,121 @@ func (n *Node) Start() error {
...
 	return nil
 }
 
 // Stop stop openbazaard
-func (n *Node) Stop() error {
+func (n *Node) Stop() {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+
+	n.stop()
```
Comment (placer14) [inline mobile/node.go]: `return n.stop()`

File: mobile/node.go [hunk 7]
```diff
@@ -404,78 +416,121 @@ func (n *Node) Start() error {
...
+}
+
+func (n *Node) stop() error {
+	//n.cancel()
```
Comment (placer14) [inline mobile/node.go]: Remove commented code.

File: mobile/node.go [hunk 8]
```diff
@@ -404,78 +416,121 @@ func (n *Node) Start() error {
...
+	n.started = false
 	return nil
 }
 
+func (n *Node) Restart() error {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+
+	if n.started {
+		n.stop()
```
Comment (placer14) [inline mobile/node.go]: Handle error returned from `n.stop()`.

File: mobile/node.go [hunk 9]
```diff
@@ -529,5 +584,5 @@ func newHTTPGateway(node *core.OpenBazaarNode, ctx commands.Context, authCookie
 	}
 
 	// Create and return an API gateway
-	return api.NewGateway(node, authCookie, manet.NetListener(gwLis), config, mainLoggingBackend, opts...)
+	return api.NewGateway(node.OpenBazaarNode, authCookie, manet.NetListener(gwLis), config, mainLoggingBackend, opts...)
```
Comment (placer14) [inline mobile/node.go]: I don't think we should be using the global node here. When we kill it or try to use any of the resources, it'll fail as soon as we start using a new node.

File: mobile/cmd/main.go (hunk 2 above)
Comment (cpacia) [reply mobile/cmd/main.go]: Doesn't seem to be anyway around this. Gofmt always moves it into this position.

File: mobile/node.go (hunk 3 above)
Comment (cpacia) [reply mobile/node.go]: It was a hack to get access to the lock from the main package which detects the sigint shutdown.

Comment (cpacia) [reply mobile/node.go]: We can fix it here but will need to be fixed in the main package before getting rid of it for the desktop.

File: mobile/node.go (hunk 9 above)
Comment (cpacia) [reply mobile/node.go]: What do you mean by global node? This is an instance of it here, not a global.

File: api/gateway.go (hunk 1 above)
Comment (cpacia) [reply api/gateway.go]: done

File: mobile/cmd/main.go (hunk 2 above)
Comment (placer14) [reply mobile/cmd/main.go]: It will if you include the line breaks as shown above.

File: mobile/node.go (hunk 9 above)
Comment (placer14) [reply mobile/node.go]: You're passing in this, which is global. (https://github.com/OpenBazaar/openbazaar-go/blob/mobile-stop/core/core.go#L105) I don't think we should be managing the running OB instance as a global variable.

Comment (placer14) [reply mobile/node.go]: Sorry, you're right. I was confusing it for what you were doing in the mobile package (which I see you've fixed.) Disregard all of this.

**Pull request merged**